src/test/fuzz/util/net.h

```diff
@@ -43,7 +43,7 @@ class FuzzedSock : public Sock
      * If `MSG_PEEK` is used, then our `Recv()` returns some random data as usual, but on the next
      * `Recv()` call we must return the same data, thus we remember it here.
      */
-    mutable std::optional<uint8_t> m_peek_data;
+    mutable std::optional<std::vector<uint8_t>> m_peek_data;
```

Comment:
from the commit message:

> FuzzedSock only supports peeking at one byte at a time, which is not
> fuzzer friendly when trying to receive long data.

I am not sure how likely is this to happen, but at least it is not ruled out by the `recv(2)` documentation: a peek read of 10 bytes returns just 1 byte and repeating the peek read of 10 bytes returns the same 1 byte over and over again. The app can never peek past the 1 byte without actually consuming/reading that byte. I guess it must be ready for this behavior from the OS or from a mocked fuzzed socket.

Reply:
I could only find the following in the docs and based the changes here on that.

```
This flag causes the receive operation to return data from
the beginning of the receive queue without removing that
data from the queue.  Thus, a subsequent receive call will
return the same data.
```

We could make it so that a peek read only returns one byte and a subsequent normal read return N bytes where the first byte corresponds to the peek byte. But I'm not sure if that would change much for our use case.

Reply:
> We could make it so that a peek read only returns one byte and a subsequent normal read return N bytes where the first byte corresponds to the peek byte.

This was the case before this PR, right?

> But I'm not sure if that would change much for our use case.

Where can I see that use case? Is there some code that does not work without the changes to `FuzzedSock::Recv()` done in this PR?

The doc you cited, in my understanding, means that what I described in the first comment of this thread is not ruled out and might happen.

Reply:
> This was the case before this PR, right?

No, a peek read before this PR would return 1 byte and a normal read following that would also only return that one byte (with the exception of sometimes padding with a bunch of zeros). This is what causes the fuzz input to never be interpreted as one continuous sequence of data, which is why mutations that insert pieces of data (e.g. from a dictionary) are ineffective.

> Where can I see that use case? Is there some code that does not work without the changes to FuzzedSock::Recv() done in this PR?

Yes the i2p harness is unable to cover most of the target code without these changes: https://github.com/bitcoin/bitcoin/pull/30230.